Sources/SourceKitLSP/Swift/CodeActions/ConvertStringConcatenationToStringInterpolation.swift

```diff
@@ -0,0 +1,184 @@
+    for component in componentsOnly {
+      guard let stringLiteral = component.as(StringLiteralExprSyntax.self) else {
+        ret.append(
+          .expressionSegment(
+            ExpressionSegmentSyntax(
+              pounds: commonPounds,
+              expressions: [
+                LabeledExprSyntax(expression: component.trimmed)
+              ]
+            )
+          )
+        )
+        continue
+      }
+
+      var segments = stringLiteral.segments
+      if let commonPounds, stringLiteral.isPoundsNotEqual(to: commonPounds) {
+        for i in segments.indices {
+          guard case let .expressionSegment(exprSegment) = segments[i] else {
+            continue
+          }
+          segments[i] = .expressionSegment(exprSegment.with(\.pounds, commonPounds))
+        }
+      }
+      ret += segments
+    }
+
+    return syntax.with(
+      \.elements,
+      [
+        ExprSyntax(
+          StringLiteralExprSyntax(
+            openingPounds: commonPounds,
+            openingQuote: .stringQuoteToken(),
+            segments: ret,
+            closingQuote: .stringQuoteToken(),
+            closingPounds: commonPounds
+          )
+        )
+      ]
+    )
+  }
+
+  /// If `exprList` is a valid string concatenation, returns 1) all elements in `exprList` with concat operators
+  /// stripped and 2) the longest pounds amongst all string literals, otherwise returns nil.
+  ///
+  /// `exprList` as a valid string concatenation must contain n >= 3 children where n is an odd number with a concat
+  /// operator `+` separating every other child, which must either be a single-line string literal or a valid
+  /// expression for string interpolation. `exprList` must also contain at least one string literal child.
+  ///
+  ///  The following is a valid string concatenation.
+  ///``` swift
+  ///"Hello " + aString + "\(1)World"
+  ///```
+  ///  The following are invalid string concatenations.
+  ///``` swift
+  ///aString + bString // no string literals
+  ///
+  ///"Hello " * aString - "World" // non `+` operators
+  ///
+  ///"""
+  ///Hello
+  ///"""
+  ///+ """
+  ///World
+  ///""" // multi-line string literals
+  ///```
+  private static func preflight(
+    exprList: ExprListSyntax
+  ) -> (componentsOnly: [ExprListSyntax.Element], longestPounds: TokenSyntax?)? {
+    var iter = exprList.makeIterator()
+    guard let first = iter.next() else {
+      return nil
+    }
+
+    var hasStringLiterals = false
+    var longestPounds: TokenSyntax?
+    var componentsOnly = [first]
+
+    if let stringLiteral = first.as(StringLiteralExprSyntax.self) {
+      guard stringLiteral.isSingleLine else {
+        return nil
+      }
+      hasStringLiterals = true
+      longestPounds = stringLiteral.openingPounds
+    }
+
+    while let concat = iter.next(), let stringComponent = iter.next() {
+      guard let concat = concat.as(BinaryOperatorExprSyntax.self),
+        concat.isConcat && !stringComponent.is(MissingExprSyntax.self)
```

Comment:
Do you think that `isConcat` improves readability here? I personally think it’s easier to understand if we just inlined `isConcat` because it gives the visual cue that we’re checking for `+` here. I personally needed to check what `isConcat` does to fully understand it.

```suggestion
        concat.operator.tokenKind == .binaryOperator("+") && !stringComponent.is(MissingExprSyntax.self)
```